examples/service-fastapi/app/security.py

```diff
@@ -0,0 +1,86 @@
+import os
+
+from fastapi import HTTPException
+from fastapi import Security
+from fastapi import status
+from fastapi.security import OAuth2AuthorizationCodeBearer
+from fastapi.security.api_key import APIKeyCookie
+from fastapi.security.api_key import APIKeyQuery
+
+from .client import get_client
+
+### For authenticated endpoints, we want to get auth from one of three ways
+### 1. "token" in the url params
+### 2. "jupyterhub-services" cookie (or config via env)
+### 3. Authorization Bearer header (with oauth to Hub support)
+auth_by_param = APIKeyQuery(name="token", auto_error=False)
+
+COOKIE_NAME = os.getenv("JUPYTERHUB_COOKIE_NAME", "jupyterhub-services")
+auth_by_cookie = APIKeyCookie(name=COOKIE_NAME, auto_error=False)
+
+if "PUBLIC_HOST" in os.environ:
+    ### When running in Docker or maybe other infrastructure,
+    ### JUPYTERHUB_API_URL is "http://jupyterhub" but we need to give
+    ### clients (swagger webpage) a link to the public url
+    auth_url = os.environ["PUBLIC_HOST"] + "/hub/api/oauth2/authorize"
+else:
+    auth_url = os.environ["JUPYTERHUB_API_URL"] + "/oauth2/authorize"
```

Comment:
jupyterhub_api_url is not generally correct. It should be calculated based on the *request* host, while the API url is only for direct service->hub communication, not user requests from the outside.

If FastAPI doesn't allow calculation of path-only authorization targets (not unusual, since 'internal' oauth is weird!), then requiring public host to always be set is probably best. Or defaulting to the default public url: `http://127.0.0.1:8000/hub/api` with a warning.

Reply:
I get turned in circles on this oauth2 stuff, it's good to talk through it.  There are four places that I deal with the auth url right now.  The problem child in this flow is Swagger, not necessarily FastAPI.

1. In jupyterhub_config.py, starting the service.
2. In security.py, setting the url that users get redirected to when they click "authorize" in the OAuth2 option in Swagger
3. In app.py, configuring Swagger UI
4. In service.py get_token route, which handles the callback from 3 and then posts to Hub to get the token

2 determines what url you get redirected to when you click on "Authorize -> OAuth2" in the swagger page.  After some testing it looks like this can just be set to a relative path `/hub/api/oauth2/authorize`.

When the redirect to the Hub login page happens, the `redirect_uri` url parameter is created by Swagger from the request host + config in 3.  Without setting `oauth_redirect_uri` in 1, you'll always get a `redirect uri mismatch` error `'<your-host>/services/fastapi/oauth_callback != '/services/fastapi/oauth_callback'`.  4 needs to match this as well.

So in summary, Swagger adds the request host to the `redirect_uri` and we need to account for that when starting the service (`jupyterhub_config.py`) and on the `/oauth2/token` callback (`service.py`) or we'll hit `redirect uri mismatch` errors.  

Would you put that warning and set the default env variable is in `jupyterhub_config.py`?

Reply:
Added a warning when `PUBLIC_HOST` env wasn't defined in `jupyterhub_config.py` and set it to `http://127.0.0.1:8000` as a default.